domain/machine/service/service.go

```diff
@@ -136,17 +153,60 @@ func (s *Service) InstanceId(ctx context.Context, machineName machine.Name) (str
 	return instanceId, nil
 }
 
-// InstanceStatus returns the cloud specific instance status for this
+// GetInstanceStatus returns the cloud specific instance status for this
 // machine.
-// If the machine is not provisioned, it returns a NotProvisionedError.
-func (s *Service) InstanceStatus(ctx context.Context, machineName machine.Name) (string, error) {
-	instanceStatus, err := s.st.InstanceStatus(ctx, machineName)
+// It returns NotFound if the machine does not exist.
+// It returns a StatusNotSet if the instance status is not set.
+// Idempotent.
+func (s *Service) GetInstanceStatus(ctx context.Context, machineName machine.Name) (status.StatusInfo, error) {
+	instanceStatus, err := s.st.GetInstanceStatus(ctx, machineName)
 	if err != nil {
-		return "", errors.Annotatef(err, "retrieving cloud instance status for machine %q", machineName)
+		return status.StatusInfo{}, errors.Annotatef(err, "retrieving cloud instance status for machine %q", machineName)
 	}
 	return instanceStatus, nil
 }
 
+// SetInstanceStatus sets the cloud specific instance status for this
+// machine.
+// It returns NotFound if the machine does not exist.
+// It returns InvalidStatus if the given status is not a known status value.
+func (s *Service) SetInstanceStatus(ctx context.Context, machineName machine.Name, status status.StatusInfo) error {
+	if !status.Status.KnownInstanceStatus() {
```

Comment:
Naive question: how can we have an unknown status here? is it because the set of accepted values for instance status is a subset of the `StatusInfo` set?

Reply:
Yeah `unknown` is a known instance status value, both per `StatusInfo` and `instance_status_value` table. Logically I agree, I don't think we'll ever get an unknown status as an argument to the SetInstanceStatus, but it is a valid status nevertheless.